spark/src/main/scala/geotrellis/spark/KeyBounds.scala

```diff
@@ -3,26 +3,93 @@ package geotrellis.spark
 import spray.json._
 import spray.json.DefaultJsonProtocol._
 
+sealed trait Bounds[+A] extends Product with Serializable {
+  def isEmpty: Boolean
+
+  def nonEmpty: Boolean = ! isEmpty
+
+  def include[B >: A](key: B)(implicit b: Boundable[B]): KeyBounds[B]
+
+  def includes[B >: A](key: B)(implicit b: Boundable[B]): Boolean
+
+  def combine[B >: A](other: Bounds[B])(implicit b: Boundable[B]): Bounds[B]
+
+  def intersect[B >: A](other: Bounds[B])(implicit b: Boundable[B]): Bounds[B]
+
+  def intersects[B >: A](other: KeyBounds[B])(implicit b: Boundable[B]): Boolean =
+    intersect(other).nonEmpty
+}
+
+case object EmptyBounds extends Bounds[Nothing] {
+  def isEmpty = true
+
+  def include[B](key: B)(implicit b: Boundable[B]): KeyBounds[B] =
+    KeyBounds(key, key)
+
+  def includes[B](key: B)(implicit b: Boundable[B]): Boolean =
+    false
+
+  def combine[B](other: Bounds[B])(implicit b: Boundable[B]): Bounds[B] =
+    other
+
+  def intersect[B](other: Bounds[B])(implicit b: Boundable[B]): Bounds[B] =
+    EmptyBounds
+}
+
 case class KeyBounds[K](
   minKey: K,
   maxKey: K
-){
-  def map(f: K => K) = KeyBounds(f(minKey), f(maxKey))
+) extends Bounds[K] {
+  def isEmpty = false
+
+  def include[B >: K](key: B)(implicit b: Boundable[B]): KeyBounds[B] =
+    KeyBounds(b.minBound(minKey, key), b.maxBound(maxKey, key))
+
+  def includes[B >: K](key: B)(implicit b: Boundable[B]): Boolean =
+    minKey == b.minBound(minKey, key) && maxKey == b.maxBound(maxKey, key)
+
+  def combine[B >: K](other: Bounds[B])(implicit b: Boundable[B]): Bounds[B] =
+    other match {
+      case KeyBounds(otherMin, otherMax) =>
+        val newMin = b.minBound(minKey, otherMin)
+        val newMax = b.maxBound(maxKey, otherMax)
+        KeyBounds(newMin, newMax)
+
+      case EmptyBounds =>
+        this
+    }
+
+  def combine(other: Bounds[K])(implicit b: Boundable[K]): KeyBounds[K] =
```

Comment:
What's the overload used for? I'm actually a bit surprised there's no ambiguous compiler errors when trying to call this, because it seems to fit the above overload type sig as well.